### DISCUSSION TITLE
Amiga 2000 keyboard sync problems

### DISCUSSION THREAD

**Discussion by Cent74** [keirf/flashfloppy#543] [Q&A]
I'm using FF 3.29 in a Gotek with rotary encoder, OLED & USB stick. I'm using the files from the HxC compatibility mode folders (HXCSDFE.CFG and AUTOBOOT.HFE).

Amiga disks on USB load well, but the keyboard usually loses sync and does only recover after reset (tested with a logic analyzer directly on the CIA pins connecting to the keyboard. Looks like Amiga does not send a complete acknowledge and keyboard already sends the next character). Only in very rare cases the keyboard syncs after some time, but this is not reproducible. Almost looks like Amiga is too busy to send a complete keyboard acknowledge in time.

When booting the file selector it takes about a minute before the keyboard works, but as soon as I boot an adf-image sync gets lost and does not recover.

Disconneting Gotek and using the original disk drive and booting real disk, the keyboard works without issues.

Any experience or ideas on this issue?

**Comment by keirf** [comment]
Not reported before. What is the Amiga model?

**Reply by Cent74** [reply]
It's an Amiga 2000 REV 4.1.

**Comment by keirf** [comment]
Is this actually FlashFloppy related at all?

**Reply by Cent74** [reply]
I tried another keyboard (modified from an Amiga 500) and it also does not work as long as Gotek is connected.

Since it works well when I use the original drive instead of Gotek, it has to be related to the Gotek in some way.

I am just not sure if this is a firmware issue, or rather something on the Gotek hardware side. But what could that be? After all, USB disk images load well with the Gotek, just the keyboard is not working properly.

**Reply by keirf** [reply]
This is a plain Gotek -- not using On Screen Display hardware, not using the keyboard for direct Gotek control (using AUTOBOOT.HFE doesn't count)?

**Reply by keirf** [reply]
That being the case, I'd suggest there's an electrical issue somehow. There is no direct interaction between the floppy header and the keyboard circuitry, so no way the Gotek can directly screw up your keyboard.

**Reply by Cent74** [reply]
Yes, its a plain Gotek.

I agree its probably something very strange on the hardware side.

Just wanted to rule out that its related to interrupt handling on the Amiga, being too busy processing data from the Gotek also incoming via interrupt. Because what I observe with the logic analyzer is that the rising edge of the acknoledge from Amiga comes too late, and keyboard starts sending the next block - staying forever out of sync. 

But I guess this cannot be flash firmware related, but rather AUTOBOOT.HFE-related, like a keyboard interrupt handler installed by AUTOBOOT or something like that)...

Anyway, if this is not a known problem then I'll continue searching and report here if I can solve it. Thanks for your help!

**Reply by keirf** [reply]
Does it always only occur if you use AUTOBOOT.HFE? What if you don't? (Remove AUTOBOOT.HFE and HXCSDFE.CFG from USB stick).

You can still mount ADF images this way, and in fact I would say navigation is easier this way using OLED rather than the file selector.

**Reply by keirf** [reply]
Since you can look at this with a scope or LA, the keyboard ISR in the file selector is the same I use in Amiga Test Kit. It handshakes by pulling KDAT low for 90 microseconds. This is just a bit more than the 85us stated in Appendix H of the Hardware Refererence Manual 2ed (page 344). I take the 85us here to be a minimum. I've never heard of anyone having problems with keyboard handling in ATK, and it's been used to test a lot of Amigas!

**Reply by Cent74** [reply]
Good point. I deleted AUTOBOOT.HFE and HXCSDFE.CFG from the USB stick, but same behaivour: booting an adf-image works, but keyboard not working.

This is what I see with the Logic analyzer:
![Keyboard not working github](https://user-images.githubusercontent.com/80603989/135980038-ae136dec-3097-4951-b337-3a7cfcb11d2f.jpg)

What we see here is a typical resync procedure (I cut off the part before in which the Kb sends "1"s to fill the CIA serial port up to 8 bits - this happens until the 8th bit is received, and then Amiga sends the acknoledge shown here on the left).

The middle part is where the Kb sends an error code to the Amiga to notify that the last byte was erroneous, but after the 8th bit Amiga pulls down Dat (to signal acknowledge) but the rising edge is missing. Then the Kb starts clocking the last character, and waits for an ack that never comes - presumably because the Amiga CIA has not received 8 bits. So reync starts again, and this happens on an on.

Very rarely though the following happens (suddenly and non reproducibly) after waiting a long time:
![Keyboard_works_again_github](https://user-images.githubusercontent.com/80603989/135981014-a6899e88-250e-42a9-86c5-df9bb910d706.jpg)

Sync is established, and the only difference is the time in the middle part between the two bytes (indicated by horizontal arrow). This is just a little bit longer than usual, and that obviously enables the CIA to detect the first clock impulse and get a complete byte. So it sends "Ack" and from now on all works stable.

**Reply by keirf** [reply]
Do you have a trace where it's working (ie. when using a real drive)? Interesting to compare. I don't think the keyboard should transmit the next character until it sees KDAT rise. Perhaps it gets confused by its own rising edge on KDAT?

**Reply by Cent74** [reply]
Yes, I'm also wondering why the keyboard just starts sending without a rising edge on KDAT. But the Amiga 500 keyboard I connected as an alterative does the same so we can rule out a problem with the keyboard.

I can do another trace with the real drive, but I cannot replicate the short interval between the two bytes if the keyboard stays in sync. Only when sync is lost it sends the erronous character within µs after the error code byte.

But I don't see how the firmware would cause this. Maybe I should check the logic levels with a scope... possibly some noise between the lines, pullups or capacitors defective etc

**Reply by keirf** [reply]
Could also be worth swapping the CIA chips round and see what happens?

**Reply by Cent74** [reply]
Just another thought: You mentioned that the keyboard ISR pulls down KDat for 90us. The traces above are from the file selector, but this looks much longer?
Also interesting to see that Int2 goes low, and so does IPL1 (indicating level 2 interrupt at the 68k). The rising edge of IP1 seems to run into the first new clock pulse, but KDat goes high much later. I think this should happen roughly at the same time.
I already swapped the CIAs (and also replaced all other chips with Amiga 500 versions) but same problem as soon as I connect the Gotek.

**Reply by keirf** [reply]
Yes it does look much longer than 90us which is weird. The ISR counts 65 CIA ticks. The CIA ticks at CPU freq / 10.

KDAT isn't going high at the end of the ISR when things go wrong because the keyboard has already begun driving KDAT itself.

I suppose the length of the IPL1 pulse is a good proxy for length of handshake. The length of those pulses being 200us seems a bit long. Then again it's just C code running on a slow 68000.

**Reply by keirf** [reply]
This is perhaps a reason to get traces on real drive and Gotek with the same disk / disk image that works on one and not the other. Then you can look for things like difference in IPL1 pulse length, handshake, etc.

**Reply by keirf** [reply]
Since this is possib'y involving analog effects you might also want to put a scope on the keyboard serial signals.

**Reply by Cent74** [reply]
Yes, I'll do three traces with the same signals as above:

1) within file selector (should replicate the graphs above)
2) booting from an adf image (not using the file selector keyboard ISR) (should also replicate the above)
3) using a real drive.
I'll connect the scope to KDat and Clk. Hope to find some time this evening...

A question regarding the ISR:
I looked at the keyb ISR some time ago and found the part where Kdat is set to output for $40 counts (this is probably the ACK counting to 65 you mentioned). Just wondering: how do you set the KDat line (=SP) low in the CIA? Doesn't that work by shifting out the serial register? I did not find any write command to the serial register though.

**Reply by keirf** [reply]
By putting the CIA's Serial Port in output mode, the CIA chip takes control of the DAT pin. Since we do not actually send a byte, the pin takes the value of the last bit sent by the CIA. Since no bytes are ever sent to the keyboard, I assume "last bit sent" must default to 0 at reset. Hence DAT is driven low until the Serial Port is flipped back into input mode. This is why the CIA loses the start of the next byte if it is sent too early: The CIA serial port is not in input mode!

**Reply by Cent74** [reply]
Okay, I did some more tests. I connected a real drive, and the keyboard worked as expected:

![1_kb_funktioniert_Drive_down_Git1](https://user-images.githubusercontent.com/80603989/136614837-75fcaf13-65ef-4036-8112-77db01261bbf.png)

Then I booted a real disk and expected (just like the last time) that the KB would continue working. But after a while it suddenly lost sync:

![geht nicht - git2](https://user-images.githubusercontent.com/80603989/136615039-fe3b5f4b-c339-4626-af7e-296c3ba49763.png)

KB and Amiga went in sync mode and never recovered. But now it's clear that it is not the Gotek.

Two things are suprising: First, the acknoledge in the upper graph (working keyboard) ist extremely long. And second, whenever I stay on the boot screen the keyboard works (tested with an Action replay module in which I can enter characters and thus test the keyboard on the boot screen). But as soon as I use the drive, sooner or later the KB will lose sync.

I checked the logic levels with a scope, and all looks fine:

![DS1Z_QuickPrint9](https://user-images.githubusercontent.com/80603989/136615443-d50e578c-6430-40c4-8a37-89961844d213.png)

My next step is to reactivate an Amiga 500 and take the same traces there. I want to see how the timing looks on a working Amiga. but apart from that I'm running out of ideas...

**Reply by Cent74** [reply]
Finally I found a solution: it was to remove the two capacitors C910 and C911 on the A2000 main board.

Thought I'd share this if anyone else has the same problem.

**Reply by keirf** [reply]
Supply decoupling capacitors? That's weird if so.

**Reply by penfold42** [reply]
i think they're on clk and data.
This was one of the C= mods to "upgrade" a rev 4.1 to rev 4.5 board

**Reply by keirf** [reply]
Makes much more sense!

**Reply by Cent74** [reply]
That's right, removing C910+911 was proposed as a mod (mainly for getting back the first keyboard character after reset). I didn't really expect it would solve my sync problem, but it did.
I found this solution in an old Amiga magazine where they said the reason was the capacitors being duplicated in both the keyboard and on the main board. Later revisions had that fixed.
